src/librustc/ty/query/config.rs

```diff
@@ -991,14 +1000,17 @@ macro_rules! impl_disk_cacheable_query(
     }
 );
 
-impl_disk_cacheable_query!(unsafety_check_result, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(borrowck, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(mir_borrowck, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(mir_const_qualif, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(check_match, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(def_symbol_name, |_| true);
-impl_disk_cacheable_query!(type_of, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(predicates_of, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(used_trait_imports, |def_id| def_id.is_local());
-impl_disk_cacheable_query!(codegen_fn_attrs, |_| true);
-impl_disk_cacheable_query!(specialization_graph_of, |_| true);
+impl_disk_cacheable_query!(mir_borrowck, |tcx, def_id| {
+    def_id.is_local() && tcx.is_closure(def_id)
+});
```

Comment:
What's the motivation here? If the result is requested (even if it's not used) then the query will be re-run. So unless we make sure that only `ensure` is used in all cases that don't need the result, this could backfire. There's at least one case where `ensure` is not used here:
https://github.com/rust-lang/rust/blob/8a0e5faec7f62e3cfd88d6625ce213d93b061305/src/librustc_driver/driver.rs#L1286

I wonder if this is worth the risk.

Reply:
This is just to avoid spending time storing the results of `mir_borrowck` for cases where it will never be used again.

Reply:
Yes, that makes sense. I wondering though if there's a way of making sure that this doesn't silently break. If somebody adds a non-`ensure()` call to `mir_borrowck` it would go mostly unnoticed. Well, I guess we would notice on perf.rlo after the fact.